internal/controllers/nmc_reconciler.go

```diff
@@ -555,104 +565,132 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
 	return errors.Join(errs...)
 }
 
-func (h *nmcReconcilerHelperImpl) UpdateNodeLabelsAndRecordEvents(ctx context.Context, nmc *kmmv1beta1.NodeModulesConfig) error {
-	node := v1.Node{}
-	if err := h.client.Get(ctx, types.NamespacedName{Name: nmc.Name}, &node); err != nil {
-		return fmt.Errorf("could not get node %s: %v", nmc.Name, err)
-	}
+type LabelPreparationHelper interface {
+	GetDeprecatedKernelModuleReadyLabels(deprecatedNodeModuleReadyLabels sets.Set[string], node v1.Node) sets.Set[string]
+	GetKernelModuleReadyLabels(nodeModuleReadyLabels sets.Set[types.NamespacedName], node v1.Node) sets.Set[types.NamespacedName]
+	GetSpecLabelsAndTheirConfigs(specLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig, nmc *kmmv1beta1.NodeModulesConfig) map[types.NamespacedName]kmmv1beta1.ModuleConfig
+	GetStatusLabelsAndTheirConfigs(statusLabels map[types.NamespacedName]kmmv1beta1.ModuleConfig, nmc *kmmv1beta1.NodeModulesConfig) map[types.NamespacedName]kmmv1beta1.ModuleConfig
+}
+type labelPreparationHelperImpl struct {
+}
 
-	// get all the kernel module ready labels of the node
-	nodeModuleReadyLabels := sets.New[types.NamespacedName]()
-	deprecatedNodeModuleReadyLabels := sets.New[string]()
+func NewLabelPreparationHelper() LabelPreparationHelper {
+	return &labelPreparationHelperImpl{}
+}
 
+func (lphi *labelPreparationHelperImpl) GetKernelModuleReadyLabels(nodeModuleReadyLabels sets.Set[types.NamespacedName], node v1.Node) sets.Set[types.NamespacedName] {
 	for label := range node.GetLabels() {
 		if ok, namespace, name := utils.IsKernelModuleReadyNodeLabel(label); ok {
 			nodeModuleReadyLabels.Insert(types.NamespacedName{Namespace: namespace, Name: name})
 		}
+	}
+	return nodeModuleReadyLabels
```

Comment:
A `set` is a map (pointer) which means that when you `Insert` to it, it get updated, therefore there is not need to also return it.

I would define the Set inside the function and return it.
A function that has no state is easier to debug and preferable IMO

Reply:
done